win32ss/user/ntuser/class.c

```diff
@@ -335,30 +335,48 @@ IntRegisterClassAtom(IN PUNICODE_STRING ClassName,
                      OUT RTL_ATOM *pAtom)
 {
     WCHAR szBuf[65];
-    PWSTR AtomName;
+    PWSTR AtomName = szBuf;
+    BOOL bFreeAtom = FALSE;
     NTSTATUS Status;
 
-    if (ClassName->Length != 0)
+    _SEH2_TRY
     {
-        /* FIXME: Don't limit to 64 characters! Use SEH when allocating memory! */
-        if (ClassName->Length / sizeof(WCHAR) >= sizeof(szBuf) / sizeof(szBuf[0]))
+        if (ClassName->Length != 0)
         {
-            EngSetLastError(ERROR_INVALID_PARAMETER);
-            return (RTL_ATOM)0;
+            if (ClassName->Length >= sizeof(szBuf))
+            {
+                AtomName = ExAllocatePoolWithTag(PagedPool,
+                                                 ClassName->Length + sizeof(UNICODE_NULL),
+                                                 TAG_STRING);
+                bFreeAtom = TRUE;
+                if (AtomName == NULL)
+                {
+                    EngSetLastError(ERROR_OUTOFMEMORY);
+                    _SEH2_YIELD(return (RTL_ATOM)0);
+                }
+            }
+
+            RtlCopyMemory(AtomName,
+                          ClassName->Buffer,
+                          ClassName->Length);
+            AtomName[ClassName->Length / sizeof(WCHAR)] = UNICODE_NULL;
+        }
+        else
+        {
+            AtomName = ClassName->Buffer;
         }
 
-        RtlCopyMemory(szBuf,
-                      ClassName->Buffer,
-                      ClassName->Length);
-        szBuf[ClassName->Length / sizeof(WCHAR)] = UNICODE_NULL;
-        AtomName = szBuf;
+        Status = RtlAddAtomToAtomTable(gAtomTable,
```

Comment:
An exception during this call shouldn't be possible, right? In that case it would seem preferable to keep it out of the SEH to avoid hiding bugs.

Reply:
We don't have a catch in the SEH, just a finally so it will propagate after freeing the buffer.

Reply:
The function should be annotated with _Maybe_raises_SEH_exception_

Reply:
I'm personally not a fan of SAL annotations that are as long as a book.

Reply:
I'm personally not a fan of the & operator as address operator. I just use it, where appropriate.